src/appleseed/renderer/modeling/postprocessingeffect/vignettepostprocessingeffect.cpp

```diff
@@ -0,0 +1,163 @@
+            const std::size_t   tile_y) const override
+        {
+            Image& image = frame.image();
+
+            assert(tile_x < image.properties().m_tile_count_x);
+            assert(tile_y < image.properties().m_tile_count_y);
+
+            Tile& tile = image.tile(tile_x, tile_y);
+            const std::size_t tile_width = tile.get_width();
+            const std::size_t tile_height = tile.get_height();
+            const Vector2u tile_offset(
+                tile_x * image.properties().m_tile_width,
+                tile_y * image.properties().m_tile_height);
+
+            for (std::size_t y = 0; y < tile_height; ++y)
+            {
+                for (std::size_t x = 0; x < tile_width; ++x)
+                {
+                    const Vector2f pixel_coord = static_cast<Vector2f>(Vector2u(x, y) + tile_offset);
+
+                    // Pixel coordinate normalized to be in the [-1, 1] range vertically.
+                    const Vector2f coord = (2.0f * pixel_coord - m_resolution) / m_normalization_factor;
+
+                    //
+                    // Port of Keijiro Takahashi's natural vignetting effect for Unity.
+                    // Recreates natural illumination falloff, which is approximated by the "cosine fourth" law of illumination falloff.
+                    //
+                    // References:
+                    //
+                    //   https://github.com/keijiro/KinoVignette
+                    //   https://en.wikipedia.org/wiki/Vignetting#Natural_vignetting
+                    //
+
+                    const float linear_radial_falloff = norm(coord) * m_intensity;
+                    const float quadratic_radial_falloff = linear_radial_falloff * linear_radial_falloff + 1.0f;
+
+                    // Inversely proportional to the fourth power of the distance from the pixel to the image center.
+                    const float inverse_biquadratic_radial_falloff = 1.0f / (quadratic_radial_falloff * quadratic_radial_falloff);
+
+                    Color4f pixel;
+                    tile.get_pixel(x, y, pixel);
+
+                    pixel.rgb() *= inverse_biquadratic_radial_falloff;
+                    tile.set_pixel(x, y, pixel);
+                }
+            }
+
+        }
+
+      private:
+        // Settings.
+        const float                     m_intensity;
+        const float                     m_anisotropy;
+
+        // Context.
+        const foundation::Vector2f      m_resolution;
+        const foundation::Vector2f      m_normalization_factor;
+    };
+}
+
+
+//
+// VignetteApplierFactory class implementation.
+//
+
+VignetteApplierFactory::VignetteApplierFactory(
+    const VignetteParams& params)
+  : m_params(params)
+{
+}
+
+void VignetteApplierFactory::release()
+{
+    delete this;
+}
+
+IEffectApplier* VignetteApplierFactory::create()
+{
+    return new VignetteApplier(m_params);
+}
+
+IEffectApplier* VignetteApplierFactory::create(
```

Comment:
Do you use both `create` methods ?

Reply:
No, only the static one (`vignettepostprocessingstage.cpp:124`), the other is required by the interface. Iirc something similar is done in the `GenericFrameRendererFactory`.

Since there's no way to enforce the definition of a static `create()` with an interface, is it better to remove it and use the non-static one? This would imply the need to create an instance of `VignetteApplierFactory` (for each `VignetteParam`).

Reply:
Removed the static one.